<a href="https://colab.research.google.com/github/rafaeldeflon/Corretor_WW/blob/main/Corretor_WW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Tratando o corpus

O corpus escolhido foi um livro de ensino médio público convertido de PDF para arquivo txt (entitulado livro.txt)

In [184]:
open("livro.txt", "r")

<_io.TextIOWrapper name='livro.txt' mode='r' encoding='UTF-8'>

In [185]:
with open("livro.txt", "r") as f:
    livro = f.read()

print(livro[:500])

                                                                                    13
  I       II       III     IV       V       VI      VII     VIII




                                                                                          Cap. 1
                                                                                          Cap. 2
  Capítulo 1




                                                                                          Cap. 3
      A Química no mundo atual e sua


In [186]:
texto_exemplo = "Olá, tudo bem?"
palavras_separadas = texto_exemplo.split()

In [187]:
print(palavras_separadas)

['Olá,', 'tudo', 'bem?']


In [188]:
print(len(palavras_separadas))

3


In [189]:
texto_exemplo = "Olá, tudo bem?"
tokens = texto_exemplo.split()

In [190]:
print(len(tokens))

3


In [191]:
print(tokens)

['Olá,', 'tudo', 'bem?']


Importando NLTK

In [192]:
import nltk 
nltk.download('punkt')
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [193]:
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)

In [194]:
def separa_palavras(lista_tokens):
    lista_palavras = []
    for token in lista_tokens:
        if token.isalpha():
            lista_palavras.append(token)
    return lista_palavras

separa_palavras(palavras_separadas)

['Olá', 'tudo', 'bem']

Utilizando o livro importado

In [195]:
lista_tokens = nltk.tokenize.word_tokenize(artigos)
lista_palavras = separa_palavras(lista_tokens)
print(f"O número de palavras é {len(lista_palavras)}")

O número de palavras é 156882


In [196]:
def normalizacao(lista_palavras):
    lista_preparada = []
    for palavra in lista_palavras:
        lista_preparada.append(palavra.lower())
    return lista_preparada

lista_preparada = normalizacao(lista_palavras)
print(lista_preparada[:5])

['i', 'ii', 'iii', 'iv', 'v']


In [197]:
stopwords = ["i", "ii", "iii", "iv", "v", "vi", "vii", "viii", "cap"]

In [198]:
def limpeza(preparada):
    lista_normalizada = []
    for palavra in lista_preparada:
      if palavra not in stopwords:
        lista_normalizada.append(palavra)
    return lista_normalizada

lista_normalizada = limpeza(lista_preparada)
print(lista_normalizada[:5])

['capítulo', 'a', 'química', 'no', 'mundo']


In [199]:
print(lista_normalizada[:5])

['capítulo', 'a', 'química', 'no', 'mundo']


In [200]:
len(lista_normalizada)

148010

Removendo palavras repetidas por meio da função 'set()'

In [201]:
set(lista_normalizada)
len(set(lista_normalizada))

11804

# Primeiro código do corretor

Primeiramente criou-se uma função capaz de inserir letras (insere_letras) em palavras digitadas erradas.

In [202]:
palavra_exemplo = "lgica"

def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    return palavras_geradas

palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

def corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key=probabilidade)
    return palavra_correta

['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'àlgica', 'álgica', 'âlgica', 'ãlgica', 'èlgica', 'élgica', 'êlgica', 'ìlgica', 'ílgica', 'îlgica', 'òlgica', 'ólgica', 'ôlgica', 'õlgica', 'ùlgica', 'úlgica', 'ûlgica', 'çlgica', 'lagica', 'lbgica', 'lcgica', 'ldgica', 'legica', 'lfgica', 'lggica', 'lhgica', 'ligica', 'ljgica', 'lkgica', 'llgica', 'lmgica', 'lngica', 'logica', 'lpgica', 'lqgica', 'lrgica', 'lsgica', 'ltgica', 'lugica', 'lvgica', 'lwgica', 'lxgica', 'lygica', 'lzgica', 'làgica', 'lágica', 'lâgica', 'lãgica', 'lègica', 'légica', 'lêgica', 'lìgica', 'lígica', 'lîgica', 'lògica', 'lógica', 'lôgica', 'lõgica', 'lùgica', 'lúgica', 'lûgica', 'lçgica', 'lgaica', 'lgbica', 'lgcica', 'lgdica', 'lgeica', 'lgfica', 'lggica', 'lghica', 'lgiica', 'lgjica', 'lgkica', 'lglica',

In [203]:
def corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key=probabilidade)
    return palavra_correta

In [204]:
frequencia = nltk.FreqDist(lista_normalizada)
frequencia.most_common(10)

[('de', 8719),
 ('a', 5576),
 ('e', 3663),
 ('o', 3392),
 ('que', 3084),
 ('da', 2555),
 ('do', 2365),
 ('em', 2056),
 ('é', 1876),
 ('os', 1654)]

In [205]:
frequencia["lógica"]

8

In [206]:
total_palavras = len(lista_normalizada)

In [207]:
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

In [208]:
frequencia = nltk.FreqDist(lista_normalizada)
total_palavras = len(lista_normalizada)
frequencia.most_common(10)

[('de', 8719),
 ('a', 5576),
 ('e', 3663),
 ('o', 3392),
 ('que', 3084),
 ('da', 2555),
 ('do', 2365),
 ('em', 2056),
 ('é', 1876),
 ('os', 1654)]

In [209]:
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

probabilidade("lógica")

5.405040199986487e-05

In [210]:
probabilidade("lagica")

0.0

In [211]:
corretor(palavra_exemplo)

'lógica'

#Código final do corretor

Após a primeira versão do corretor, que corrige palavras que faltam uma letra (insere_letras), adiciona-se novas funções para outros tipos de erros de grafia: para letras escritas a mais (deletando_caracteres), para letras que não são da palavra trocadas (troca_letra) e letras escritas em posições trocadas da mesma palavra (inverte_letra).

In [212]:
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzáàãâéèêíìóòõôúùûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

def inverte_letra(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1] + D[0] + D[2:])
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    return palavras_geradas

palavra_exemplo = "lígica"
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alígica', 'blígica', 'clígica', 'dlígica', 'elígica', 'flígica', 'glígica', 'hlígica', 'ilígica', 'jlígica', 'klígica', 'llígica', 'mlígica', 'nlígica', 'olígica', 'plígica', 'qlígica', 'rlígica', 'slígica', 'tlígica', 'ulígica', 'vlígica', 'wlígica', 'xlígica', 'ylígica', 'zlígica', 'àlígica', 'álígica', 'âlígica', 'ãlígica', 'èlígica', 'élígica', 'êlígica', 'ìlígica', 'ílígica', 'îlígica', 'òlígica', 'ólígica', 'ôlígica', 'õlígica', 'ùlígica', 'úlígica', 'ûlígica', 'çlígica', 'laígica', 'lbígica', 'lcígica', 'ldígica', 'leígica', 'lfígica', 'lgígica', 'lhígica', 'liígica', 'ljígica', 'lkígica', 'llígica', 'lmígica', 'lnígica', 'loígica', 'lpígica', 'lqígica', 'lrígica', 'lsígica', 'ltígica', 'luígica', 'lvígica', 'lwígica', 'lxígica', 'lyígica', 'lzígica', 'làígica', 'láígica', 'lâígica', 'lãígica', 'lèígica', 'léígica', 'lêígica', 'lìígica', 'líígica', 'lîígica', 'lòígica', 'lóígica', 'lôígica', 'lõígica', 'lùígica', 'lúígica', 'lûígica', 'lçígica', 'líagica', 'líbgica', 'lícgica'

#Avaliação do corretor

Para avaliar o corretor contruiu-se uma função que testará o programa com uma lista de palavras erradas.

In [213]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []

    return lista_palavras_teste

def avaliador(testes):
    print("taxa de acerto")

In [214]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r")
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

def avaliador(testes):
    print("taxa de acerto")

In [215]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r")
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

lista_teste = cria_dados_teste("palavras.txt")
lista_teste

[('podemos', 'pyodemos'),
 ('esse', 'esje'),
 ('já', 'jrá'),
 ('nosso', 'nossov'),
 ('são', 'sãêo'),
 ('dos', 'dosa'),
 ('muito', 'muifo'),
 ('imagem', 'iômagem'),
 ('sua', 'ósua'),
 ('também', 'tambéùm'),
 ('ele', 'eme'),
 ('fazer', 'èazer'),
 ('temos', 'temfs'),
 ('essa', 'eàssa'),
 ('quando', 'quaôdo'),
 ('vamos', 'vamvos'),
 ('sobre', 'hsobre'),
 ('java', 'sjava'),
 ('das', 'daõs'),
 ('agora', 'agorah'),
 ('está', 'eòtá'),
 ('cada', 'céda'),
 ('mesmo', 'zmesmo'),
 ('nos', 'noâ'),
 ('forma', 'fobma'),
 ('seja', 'sejéa'),
 ('então', 'enêão'),
 ('criar', 'èriar'),
 ('código', 'cóeigo'),
 ('caso', 'casío'),
 ('exemplo', 'áexemplo'),
 ('tem', 'tĩem'),
 ('usuário', 'usuárôio'),
 ('dados', 'dfados'),
 ('python', 'pgthon'),
 ('nossa', 'nossah'),
 ('além', 'alémè'),
 ('assim', 'asõim'),
 ('ter', 'teb'),
 ('até', 'atĩ'),
 ('bem', 'âem'),
 ('design', 'desigen'),
 ('trabalho', 'trabalàho'),
 ('foi', 'foo'),
 ('apenas', 'apenaũ'),
 ('empresa', 'empresà'),
 ('valor', 'valíor'),
 ('será', 'serr')

In [216]:
def avaliador (testes):
    print("taxa de acerto")

In [217]:
def avaliador (testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras")

avaliador(lista_teste)

65.05% de 186 palavras


O resultado final é de uma eficiÊncia de aproximadamente 65% de acerto de correção para o corretor corrigindo 186 palavras aleatórias. 

É um número relativamente baixo considerando-se que são palavras completamente aleatórias de origens e áreas do conhecimento distintas.

Contudo, considerando-se que trata-se de um corretor que utiliza um corpus construído a partir de um livro de química, um tema bastante específico, não é uma eficiência tão baixa, ainda mais levando-se em conta que a finalidade de uso deste programa seria para correções de textos de temas desta área.